[flang][NFC] Refactor to remove .inc file containing shared code

tarunprabhu opened the pull request:
Remove flang/include/flang/Tools/CLOptions.inc - which was included as is in - several places. Move the code in it to header and source files which are used used in the "standard" way. Some minor cleanup such as removing trailing whitespace and excessive newlines and reordering entries alphabetically for files that were modified along the way. Update the documentation that referenced CLOptions.inc.

------------------------

Notes for reviewers:

This is in response to #102975 being reverted. The underlying issue was `flang/include/flang/Tools/CLOptions.inc` being included directly in several places. That file contains command-line options and function definitions which results in the duplicate definition errors if the file is included in more than one place in an executable/library.

IMO, including files containing code causes issues in the long run as it removes the ability to link against certain libraries. This PR moves the code in `CLOptions.inc` into `.h` and `.cpp` files which can then be included and linked in the "standard" way.

The contents of `CLOptions.inc` are split across two places since the contents include both command line options and functions to build FIR pass pipelines which don't seem like they should belong together anyway.

The command line options were moved to `Common/` since it seemed a "reasonable" place for them. 

The FIR pass pipeline code was moved to a new `Passes` directory in `flang/lib/Optimizer`. It was not possible to put the code in, for instance, `Common/`, `Support/` or `Optimizer/Transforms/` because that resulted in circular dependencies. The name `Passes` was chosen because the code there is functionally similar to some code in `llvm/Passes`.

Commit:
[flang] Refactor to remove .inc file containing shared code

Remove flang/include/flang/Tools/CLOptions.inc - which was included as is in -
several places. Move the code in it to header and source files which are used
used in the "standard" way. Some minor cleanup such as removing trailing
whitespace and excessive newlines and reordering entries alphabetically for
files that were modified along the way. Update the documentation that
referenced CLOptions.inc.

vzakhari commented on flang/lib/Common/CMakeLists.txt:
```diff
@@ -44,5 +44,6 @@ add_flang_library(FortranCommon
   ${version_inc}
 
   LINK_COMPONENTS
+  Passes
```
Is this dependency really needed?

tarunprabhu replied:
Yes. The default optimization level is `llvm::OptimizationLevel::O0`. This is a class instance which is defined in `libLLVMPasses`. Without it, one gets an undefined reference error when `-DBUILD_SHARED_LIBS=ON`

vzakhari commented on flang/lib/Optimizer/Passes/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+add_flang_library(FIRPasses
```
I would rather avoid `FIR` here, because these may be `HLFIR` passes as well.  To avoid any confusion, I think, `FlangPasses` might be a better name.

tarunprabhu replied:
Done. I have used `flangPasses` (note the lower case 'f') to be consistent with some of other `libflang` libraries.

vzakhari commented on flang/lib/Common/CMakeLists.txt:
```diff
@@ -34,8 +34,8 @@ if(FLANG_VENDOR)
 
-
 add_flang_library(FortranCommon
+  CommandLineOpts.cpp
```
It does not feel like it belongs to `FortranCommon`.  Can we put it into the same component library as `Pipelines.cpp`?

tarunprabhu replied:
Done

vzakhari reviewed (commented):
Thank you for the clean-up, Tarun!

jeanPerier commented on flang/include/flang/Common/CommandLineOpts.h:
```diff
@@ -0,0 +1,60 @@
+//===-- CommandLineOptions.h -- shared command line options -----*- C++ -*-===//
```
I agree with Slava that this does not belong to Common, CommandLineOpts, putting it in the same library as  Pipelines.cpp makes sense to me.

nit: "CommandLineOptions.h" -> "CommandLineOpts.h"

tarunprabhu replied:
Done.

jeanPerier reviewed (commented):
Thanks Tarun

tblah reviewed (commented):
Thanks for the cleanup! It looks good to me with the comments already mentioned

Commit:
Address reviewer comments.

Move CommandLineOpts from flang/Common to flang/Optimizer/Passes
Rename the library to flangPasses from FIRPasses.
Expose the API from namespace fir that was available with the old CLOptions.inc
file.
Fix indentation in CMakeLists.txt.

tarunprabhu commented on flang/lib/Optimizer/Passes/Pipelines.cpp:
```diff
@@ -0,0 +1,330 @@
+  pm.addPass(mlir::createCanonicalizerPass(config));
+}
+
+static void addCfgConversionPass(mlir::PassManager &pm,
```
I have exposed everything in `namespace fir`. If you need the functions in the anonymous namespace as well, let me know and I will expose that too.

vzakhari replied:
Thank you very much, Tarun! Yes, can you please also export the `addNestedPassToAllTopLevelOperations*` helpers? The `namespace fir` works well.

tarunprabhu replied:
Done. I moved everything from the anonymous namespace to the FIR namespace. For the non-templated functions, I leaving them in an anonymous namespace resulted in "unused function" warnings. My understand is that the use of functions in that way could result in ODR violations, so I moved everything, including the templated functions, into the FIR namespace.

Commit:
Run clang-format

Commit:
Expose the addNestedPassToAll* API's as well.
Move those into the fir namespace since leaving them in the anonymous namespace
may result in ODR violations.

vzakhari reviewed (approved):
LGTM. Thank you, Tarun!

tarunprabhu commented:
Thank you all for the reviews.
